Fix Issue #53

allenxwang opened the pull request:
Fix Issue #53.

Created new JUnit test for POST with content.

Commit:
Fix #53

Commit:
Merge branch 'master' of github.com:Netflix/RxNetty

Commit:
Added missing files from last commit.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/ClientRequestResponseConverter.java:
```diff
@@ -96,8 +96,15 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
                 allWritesListener.listen(ctx.write(rxRequest.getNettyRequest()));
-                if (rxRequest.hasRawContentSource()) {
-                    RawContentSource<?> rawContentSource = rxRequest.getRawContentSource();
+                ContentSource<?> contentSource = null;
```
null initialization is redundant.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/ClientRequestResponseConverter.java:
```diff
@@ -96,8 +96,15 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
-                if (rxRequest.hasRawContentSource()) {
-                    RawContentSource<?> rawContentSource = rxRequest.getRawContentSource();
+                ContentSource<?> contentSource = null;
+                if (rxRequest.getContentFactory() != null) {
```
Would it be better if we do not expose the factory from HttpRequest? Ideally this code must just do request.getContentSource() and then request.isRaw(contentSource) to handle raw of otherwise content source.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/ClientRequestResponseConverter.java:
```diff
@@ -96,8 +96,15 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+                }
+                if (contentSource instanceof RawContentSource) {
+                    @SuppressWarnings("rawtypes")
+                    RawContentSource<?> rawContentSource = (RawContentSource) contentSource;
```
You don't need the wildcard <?> here.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/ContentFactory.java:
```diff
@@ -0,0 +1,29 @@
+ * @param <T>
+ */
+public interface ContentFactory<T, R extends ContentSource<T>> {
+    public R newContentSource();
```
keyword public is redundant for interface methods.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/RepeatableContentHttpRequest.java:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+class RepeatableContentFactory<T> implements ContentFactory<T, ReplayingContentSource<T>> {
```
I guess you wanted to make these classes as inner classes and not multiple top level classes in the same file, rite?

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/HttpRequest.java:
```diff
@@ -33,14 +34,15 @@
-    private ContentSource<T> contentSource;
+    protected ContentSource<T> contentSource;
+    protected ContentFactory<T, ?> contentFactory;
+    private AtomicBoolean contentSet = new AtomicBoolean();
```
Since HttpRequest is not required to be threadsafe, you can do with a boolean instead of an AtomicBoolean. OTOH, it should be a final, if its an AtomicBoolean

allenxwang replied:
I prefer the AtomicBoolean as it makes the check and set code clean.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/ContentFactory.java:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @param <T>
+ */
+public interface ContentFactory<T, R extends ContentSource<T>> {
```
Will ContentSourceFactory be more indicative of what it is?

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/RepeatableContentHttpRequest.java:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+class RecordingRawContentSource<T> extends ReplayingContentSource<T> implements RawContentSource<T> {
```
If you are a fan of consistent naming, it should be called ReplayingRawContentSource :)

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/RepeatableContentHttpRequest.java:
```diff
@@ -0,0 +1,115 @@
+
+class ReplayingContentSource<T> implements ContentSource<T> {
+    private ContentSource<T> original;
+    private LinkedBlockingQueue<T> recorded = new LinkedBlockingQueue<T>();
```
Since, this is not supposed to be threadsafe, you may just use a list.

allenxwang replied:
Actually using a non concurrent list will fail the JUnit test `RepeatableRequestTest`. The reason is that even in the same thread, the content of the list may be altered after its iterator is created, causing ConcurrentModificationException thrown.

NiteshKant replied:
I see, then, the use of iterator is not correct. Looking at the code, it will be difficult for anyone to understand as to why this is a concurrent data structure when the class does not need to be threadsafe.
I guess all you need is an arraylist and the current index.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/RepeatableContentHttpRequest.java:
```diff
@@ -0,0 +1,115 @@
+
+class RepeatableContentFactory<T> implements ContentFactory<T, ReplayingContentSource<T>> {
+
+    private ReplayingContentSource<T> recordingSource;
```
For the sake of consistency, this should be called replayingSource.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/RepeatableContentHttpRequest.java:
```diff
@@ -0,0 +1,115 @@
+    
+    @Override
+    public boolean hasNext() {
+        if (iteratorOfRecorded.hasNext()) {
```
can be replaced by iteratorOfRecorded.hasNext() || original.hasNext()

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/RepeatableContentHttpRequest.java:
```diff
@@ -0,0 +1,115 @@
+
+class RecordingRawContentSource<T> extends ReplayingContentSource<T> implements RawContentSource<T> {
+    
+    private RawContentSource<T> original;
```
Since, the only thing you need from the original source here is the transformer, you may just store that & it will be nice to make this variable final.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/HttpRequest.java:
```diff
@@ -33,14 +34,15 @@
     private final HttpRequestHeaders headers;
-    private RawContentSource<?> rawContentSource;
-    private ContentSource<T> contentSource;
+    protected ContentSource<T> contentSource;
```
Thinking about this more, I think it is better to only have a contentFactory and a rawContentFactory here and wrap all withContent() as a factory. That way you don't have to check for content already set & would make this code simpler.
It will be imperative to have both contentFactory & rawFactory as the rawFactory type would be a wild card &lt;?&gt; whereas ContentFactory will be &lt;T&gt;

allenxwang replied:
I thought we would need to check for content already set anyway, right? Otherwise it leads to unclear behavior to user when they can set multiple content with either withContent() or withContentFactory().

allenxwang replied:
`RepeatableContentHttpRequest` needs to access the content source of the wrapped request, and if there is no source but only factory it will not wrap it again with `RepeatableContentFactory`. So removing contentSource would make its logic complicated.

Another solution would be wrap all withContent() calls with withContentSource() or withRawContentSource(). And remove the access to factory, thus moving all the logic to the getContentSource() the only way to get the source.

NiteshKant replied:
So, what I am saying is that you will only have two instance variables, viz.,

ContentFactory<T, ContentSource<T>> contentFactory;
ContentFactory<?, RawContentSource<?>> rawContentFactory;

Now, methods withContent() will set the contentFactory as a ContentFactory implementation that just returns the same instance all the time. Methods withRawContent() will set the rawContentFactory. On top of it you will have methods to set the contentFactory and rawContentFactory.

With the above, you do not need to check if the content/factory is set or not as you will always overwrite the factory.

The other package-private methods getContentSource(), getRawContentSource() and hasRawContent() remains the same and so does the usage of these.

getContentSource() will return contentFactory.newSource()
getRawContentSource() will return rawFactory.newSource()
hasRawContent() will return null != rawFactory

Does this make sense?

NiteshKant replied:
Sorry I did not see your other note. Since contentSource is not final, I would say its safer to not make it protected. Instead, you can just call withContentFactory() from RepeatableContentHttpRequest

" thus moving all the logic to the getContentSource() the only way to get the source."
Yes thats what I was referring to.

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/HttpRequest.java:
```diff
@@ -33,14 +34,15 @@
-    private RawContentSource<?> rawContentSource;
-    private ContentSource<T> contentSource;
+    protected ContentSource<T> contentSource;
+    protected ContentFactory<T, ?> contentFactory;
```
The appropriate generic here is ContentFactory&lt;T, ? extends ContentSource&lt;T&gt;&gt;

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/HttpRequest.java:
```diff
@@ -130,12 +146,7 @@ public String getUri() {
-
-    boolean hasRawContentSource() {
-        return null != rawContentSource;
+    final ContentFactory<T, ?> getContentFactory() {
```
I think we should not expose the factory. getContentSource() should just do a factory.newSource()

allenxwang replied:
What's the difference? These are package private APIs anyway and more like implementation details.

allenxwang replied:
OK, thinking about this again I realized that doing this will encapsulate the logic in one class and help our own APIs that uses HttpRequest does not have to repeat the same check whether it is a factory or source. Is that the intention?

NiteshKant replied:
Yes thats the intention.

NiteshKant commented:
Apologies, I had to merge my pull request, so you would have to merge.

Commit:
Fix issues identified in the code review.

Commit:
Fix Issue #58

Merge branch 'master' of github.com:Netflix/RxNetty

Conflicts:
	src/main/java/io/reactivex/netty/protocol/http/client/HttpRequest.java

NiteshKant commented on src/main/java/io/reactivex/netty/protocol/http/client/HttpRequest.java:
```diff
@@ -79,17 +106,46 @@
 
     public HttpRequest<T> withContentSource(ContentSource<T> contentSource) {
-        this.contentSource = contentSource;
+        if (!contentSet.compareAndSet(false, true)) {
```
Since, we are overwriting the content factory, why do we need this check? 
The con is that it limits overriding the content later, in case it is to be reused.

allenxwang replied:
My intention is that only one method from withContent*() can called. User should create new Request object to set a different content source/factory.

NiteshKant replied:
Sure but that constraint is not adding value to the code in terms of state
management, code clarity, etc. OTOH, it is making it difficult to reuse.
So, it seems to me that it is not required. Specially, since this is a
public API, I would recommend removing such a constraint.

On Friday, February 28, 2014, allenxwang notifications@github.com wrote:

> In src/main/java/io/reactivex/netty/protocol/http/client/HttpRequest.java:
> 
> > @@ -79,17 +106,46 @@
> >      }
> > 
> > ```
> >  public HttpRequest<T> withContentSource(ContentSource<T> contentSource) {
> > ```
> > -        this.contentSource = contentSource;
> > -        if (!contentSet.compareAndSet(false, true)) {
> 
> My intention is that only one method from withContent*() can called. User
> should create new Request object to set a different content source/factory.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/Netflix/RxNetty/pull/57/files#r10188652
> .

allenxwang replied:
Fair point. I have removed such checking in pull request

https://github.com/Netflix/RxNetty/pull/60

I will add more questions/comments on the issue itself.
